File: 02-逆向工具/decrypt.js
const crypto = require('crypto');
const fs = require('fs');

const inputFile = process.argv[2] || '__APP__.wxapkg';
const outputFile = process.argv[3] || '__APP__.decrypted.wxapkg';
const wxid = process.argv[4] || 'wxca87659badc84f9c';

const data = fs.readFileSync(inputFile);

// Check V1MMWX header
const header = data.slice(0, 6).toString('utf-8');
if (header !== 'V1MMWX') {
    console.error('Not a V1MMWX encrypted wxapkg file');
    process.exit(1);
}

console.log(`Decrypting ${inputFile} with wxid=${wxid}...`);

// Derive key using PBKDF2
const salt = 'saltiest';
const iv = 'the iv: 16 bytes';
const key = crypto.pbkdf2Sync(wxid, salt, 1000, 32, 'sha1');

// The encrypted data starts at offset 6, first 1024 bytes are AES encrypted
const encryptedPart = data.slice(6, 6 + 1024);
const restPart = data.slice(6 + 1024);

// Decrypt first 1024 bytes
const decipher = crypto.createDecipheriv('aes-256-cbc', key, iv);
decipher.setAutoPadding(false);
let decrypted = decipher.update(encryptedPart);
decrypted = Buffer.concat([decrypted, decipher.final()]);

// XOR key: wxid[len-2] or 0x66 as fallback
const xorByte = wxid.length >= 2 ? wxid.charCodeAt(wxid.length - 2) : 0x66;
console.log(`XOR byte: 0x${xorByte.toString(16)} ('${String.fromCharCode(xorByte)}')`);

const restDecrypted = Buffer.alloc(restPart.length);
for (let i = 0; i < restPart.length; i++) {
    restDecrypted[i] = restPart[i] ^ xorByte;
}

// Combine: take first 1023 bytes of AES decrypted (drop last padding byte) + XOR'd rest
const result = Buffer.concat([decrypted.slice(0, 1023), restDecrypted]);
fs.writeFileSync(outputFile, result);

console.log(`Decrypted! Output: ${outputFile} (${result.length} bytes)`);
console.log(`First bytes: ${result.slice(0, 4).toString('hex')}`);

if (result[0] === 0xBE) {
    console.log('Valid wxapkg header detected!');
} else {
    console.log('Header byte is 0x' + result[0].toString(16) + ', expected 0xBE.');
}

File: 02-逆向工具/decrypt.py
import hashlib
import sys
from Crypto.Cipher import AES

def decrypt_wxapkg(input_file, output_file, wxid):
    with open(input_file, 'rb') as f:
        data = f.read()

    # Check header
    if data[:6] != b'V1MMWX':
        print('Not a V1MMWX encrypted file')
        return False

    print(f'Decrypting {input_file} with wxid={wxid}...')
    print(f'File size: {len(data)} bytes')

    # Derive key
    dk = hashlib.pbkdf2_hmac('sha1', wxid.encode('utf-8'), b'saltiest', 1000, 32)
    print(f'Derived key: {dk.hex()}')
    print(f'Key first byte: 0x{dk[0]:02x}')

    iv = b'the iv: 16 bytes'

    # AES decrypt first 1024 bytes after 6-byte header
    encrypted_part = data[6:6+1024]
    cipher = AES.new(dk, AES.MODE_CBC, iv)
    decrypted_first = cipher.decrypt(encrypted_part)

    # XOR the rest - try with first byte of derived key
    xor_byte = dk[0]
    rest = data[6+1024:]
    decrypted_rest = bytes([b ^ xor_byte for b in rest])

    result = decrypted_first + decrypted_rest

    # Check validity
    print(f'Result first byte: 0x{result[0]:02x}')
    if result[0] == 0xBE:
        print('Valid wxapkg header!')
    else:
        print('Invalid header!')

    # Debug: show bytes around 1024 boundary
    print(f'\nBytes around 1024 boundary:')
    print(f'  Decrypted [1020:1024]: {decrypted_first[1020:1024].hex()} = {repr(decrypted_first[1020:1024])}')
    print(f'  XOR rest  [0:4]:      {decrypted_rest[0:4].hex()} = {repr(decrypted_rest[0:4])}')
    print(f'  Raw rest  [0:4]:      {rest[0:4].hex()} = {repr(rest[0:4])}')

    # Try to find correct XOR byte by checking what makes valid ASCII around boundary
    # We know the data around 1024 should be part of a file path
    print(f'\nTrying different XOR bytes around boundary:')
    for test_byte in range(256):
        test_result = bytes([b ^ test_byte for b in rest[:20]])
        try:
            text = test_result.decode('ascii')
            if all(c.isprintable() or c in '\r\n\t' for c in text):
                if any(c in text for c in 'abcdefghijklmnopqrstuvwxyz/-_.'):
                    print(f'  XOR 0x{test_byte:02x}: {repr(text)}')
        except:
            pass

    with open(output_file, 'wb') as f:
        f.write(result)

    print(f'\nSaved to {output_file}')
    return True

if __name__ == '__main__':
    input_file = sys.argv[1] if len(sys.argv) > 1 else '__APP__.wxapkg'
    output_file = sys.argv[2] if len(sys.argv) > 2 else '__APP__.decrypted.wxapkg'
    wxid = sys.argv[3] if len(sys.argv) > 3 else 'wxca87659badc84f9c'
    decrypt_wxapkg(input_file, output_file, wxid)
